### ISSUE TITLE
Category following

### ISSUE THREAD

**Issue by tburry** [vanilla/vanilla#6026]
We want the ability to follow categories to improve the UX on communities with many categories. This is not the same as Vanilla's current "unfollow" functionality, but will replace that.

## Requirements

- Users can follow or unfollow a category. Only signed in users can follow categories.
- The recent discussions page will have the ability to filter by just followed categories.
- The categories page will have the ability to filter by just followed categories. This will be a flat style category display.
- When filtering discussions or categories by all or followed, the user's selection should be saved so that going back to the URL shows their last selection. In this way users are customizing their UI.
- Users will be able to receive notifications for their followed categories which will be a more scalable way to implement advanced notifications. (message queue required)
- Category selection dropdowns should eventually display followed categories first.
- You can only follow a category with displayAs: discussions
- The dropdown in category lists and tables should have Follow/Unfollow options, replacing the "Mute" options.
- The category following functionality should be conditional based on the Vanilla.EnableCategoryFollowing config setting. There should be a way to toggle this setting in the dashboard.

## Design

- [Follow button](https://drive.google.com/drive/folders/12HWpLjOyAVvns5U2e6IJ2fIHkZ0uNmi_)
- View filter
- Message when the user hasn't followed any categories

## Implementation

- Implement `PUT /api/v2/categories/:id/follow` and `CategoryModel->follow()` (#6397).
- Add a filter for followed categories to `GET /api/v2/discussions` and `DiscussionModel->getWhere()` (#6398).
- Add a filter for followed categories to `GET /api/v2/categories` and `CategoryModel->getWhere()` (#6399).
- Add a filter for followed categories to `GET /api/v2/discussions/search`. The followed categories will be queried from the CategoryModel and passed into Sphinx (vanilla/internal#1346).
- Add following querystring options to /discussions and /categories (#6400).
- Add the follow/unfollow button to category page (#6401).
- Add filter dropdowns to the application to switch followed categories and all. The new base theme has this functionality planned already, but the current base theme is going to need some work later and ideation will need to be updated at that time. (#6402)
- Add config options to save the users previous selection so going to /categories or /discussions will select the current config option. (#6403)
- Add followed categories option to categories dropdowns. This will require the work done replacing the category select with a scalable category search component similar to the one in the Hootsuite app. (#6405)
- Add a followed categories option to advanced search. (vanilla/internal#1347)
- The view filter is now a core feature and must be integrated with ideation. (vanilla/internal#1348)

## Optimization

- Lets add a config cap on the max number of followed categories, defaulting to 100. The max should be checked when following a new category.
- Permissions to view a category should be checked when following a category, but not when querying followed discussions. If this generates support issues then we can loop through the returned discussions and remove ones without permissions which will be a general-purpose optimization longer term.

**Comment by tburry** [comment]
I'm hoping to get started on this 2018 ASAP. The API endpoints can be implemented without the UI functionality ahead of time.

**Comment by DaazKu** [comment]
Is that implying a rework of CategoryWatch?

**Comment by tburry** [comment]
That's right.

**Comment by Bparm** [comment]
It would be great if this feature affected our Recent Discussions Module as well.

**Comment by Charrondev** [comment]
We should get this into the roadmap repo as well.

**Comment by angeloudy** [comment]
Is there a task to add email notifications to this feature (as per the original notifications)? I'm trying to understand how this new feature intersects with the existing Advanced Notifications functionality. Will it completely replace it?

**Comment by angeloudy** [comment]
@linc Hi there! I notice you closed this task just now. Does that mean you don't have short term plans to implement the replacement of notifications with the new follow feature, or is the discussion happening somewhere else?

My company has a bit of a budget to implement this for our own needs, so we'd like to work with you to do this in the right way and contribute back to the project. Our goals are:

1. Add an attribute to the follow feature for whether that category sends emails. I'm thinking this might be an enum with choices like 'digest', 'full', 'summary', 'comments' etc. This could be an attribute per follow object (like advanced notifications is today), or just a global settings per user which applies to all follow.
2. Enhance the email template to send the actual content of the message rather than just the subject if "full" is chosen.
3. Wrap all the above in an API which where we can bulk update lots of users
4. Add an unsubscribe link to the bottom of the email which doesn't require a login

In short, we want to bring some mailing list features to vanilla

**Comment by linc** [comment]
HI @angeloudy, thanks for reaching out. I apologize I missed your comment as I was cleaning up our to-do list. We do still intend to implement notifications, but we're blocked internally by needing our message queue technology fully deployed. 

PHP's synchronous request handling makes mailing list features untenable for most open source implementations because it needs to send all emails inline with a page request, leading to timeouts when you have hundreds of notifications being generated - and that's even with offloading the email sending to a third-party service. 

I've seen other projects sidestep this limitation with batching tasks via a pseudo-cron job (basically a poor man's message queue in the application itself). I think that would had a ton of complexity to Vanilla that we're not prepared to handle. Do you have any other ideas about how this could work from an architectural perspective? We've researched this a good bit over the years and came down on the side of "message queue or bust".

It's currently possible to hook into Vanilla to insert the full message content to notifications via addon.

We have some stub issues on this repo about notifications endpoints we plan to build on the new API v2. That is definitely how we'd move forward with the message queue-based implementation, and we'd like to make all notifications happen via the API as we continue to strive for 100% platform availability via API. I think there are some larger-scope changes we want to fix in the activity & notifications framework before we make these enhancements. Tacking API endpoints onto what we have now would probably exacerbate some issues we currently have maintaining that system. (I'm not being vague on purpose, I'm just not able to summon those details off the top of my head. It's been a while since I went digging in there and unfortunately the issues I found were poorly detailed.)

Allowing "unsubscribe" with a singe-use token to sidestep login is a great feature we definitely want to do. I think that's best done as a core enhancement to also apply to all existing notifications. I don't think it needs to be tied to another project explicitly. I'm happy to review a draft scope or acceptance criteria for that feature if you want to move forward on it.

**Comment by ari** [comment]
Hi @linc 

First, I'm not a PHP coder. I once dabbled in drupal 10 years ago. But that's about it. But the general architectural issue is probably the same in any language which has only basic threading support whether it be Ruby or PHP. Some thoughts:

* the DB is a good place to keep the queue. Until it scales too far, but that might be a problem for the future. Abstract everything from the DB itself and later on Redis or something else can be dropped in to manage the queue. Personally I reckon a queue of up to 100,000 items is fine in the DB.
* thread the insertion into the queue with something async like https://github.com/dazzle-php/mysql and you'll not get performance issues when creating new posts in the UI
* decide whether the queue contains rendered messages or just pointers. If they are just pointers then more checking is needed on processing the queue for messages which are now deleted or edited.
* don't overuse transactions when adding/deleting from the queue. Only lock what's really needed to avoid contention. Queue objects should be considered immutable in the DB.
* if the work of rendering is done when inserting into the queue, then perhaps only one de-queing thread is needed since its only job is to open SMTP connections
* how to run the server de-queing thread? Frankly I have no idea... phabricator does it with a command line PHD daemon you need to launch separately. Drupal uses (or did) a /cron.php url you call regularly.

As for unsubscribe token, at a basic level that's a hash of salt + username. But I've written a more sophisticated system a few years ago which signs the username + date like 'ari:02062018:672t5gwfhjrgqyrfFHGER' with a key stored on the server. That then lets you authenticate the token but also have it expire. Obfuscate the user/date by base64 encoding it first for extra points.
